test/library/encryption/helpers.go

```diff
@@ -36,55 +86,194 @@ func GetClients(t testing.TB) (EtcdClient, configv1client.APIServerInterface, op
 	kubeClient := kubernetes.NewForConfigOrDie(kubeConfig)
 	etcdClient := NewEtcdClient(kubeClient)
 
-	operatorClient, err := operatorclient.NewForConfig(kubeConfig)
+	operatorClient, err := operatorv1client.NewForConfig(kubeConfig)
 	require.NoError(t, err)
 
-	return etcdClient, apiServerConfigClient, operatorClient.KubeAPIServers()
+	return ClientSet{Etcd: etcdClient, ApiServerConfig: apiServerConfigClient, Operator: operatorClient.KubeAPIServers(), Kube: kubeClient}
 }
 
-// WaitForOperatorAndMigrationControllerAvailableNotProgressingNotDegraded waits for the operator and encryption migration controller to report status as active not progressing, and not failing
-func WaitForOperatorAndMigrationControllerAvailableNotProgressingNotDegraded(t testing.TB, operatorClient operatorclient.KubeAPIServerInterface) {
+func WaitForEncryptionKeyBasedOn(t testing.TB, kubeClient kubernetes.Interface, prevKeyMeta EncryptionKeyMeta, encryptionType configv1.EncryptionType) {
+	encryptionMode := string(encryptionType)
+	if encryptionMode == "" {
+		encryptionMode = defaultEncryptionMode
+	}
+	if len(prevKeyMeta.Name) == 0 {
+		prevKeyMeta.Mode = defaultEncryptionMode
+	}
+
+	if prevKeyMeta.Mode == encryptionMode {
+		waitForNoNewEncryptionKey(t, kubeClient, prevKeyMeta)
+		return
+	}
+	WaitForNextMigratedKey(t, kubeClient, prevKeyMeta)
+}
+
+func waitForNoNewEncryptionKey(t testing.TB, kubeClient kubernetes.Interface, prevKeyMeta EncryptionKeyMeta) {
 	t.Helper()
-	t.Log("Waiting for Operator and Migration Controller to be: OperatorAvailable: true, OperatorProgressing: false, MigrationProgressing: false, MigrationDegraded: false, Done: true (Migration and Operator not progressing for > 1 min)")
-	// given that a happy path scenario needs to roll out at least 3 revision (each taking ~10 min)
-	// the loop below will need at least 30 min to be satisfied.
-	// we sleep here 5 minutes before we enter the loop to make sure that the condition
-	// will not be satisfied immediately
-	// TODO: consider watching other conditions and events - things like that indicate we make progress
-	time.Sleep(5 * time.Minute)
-
-	observedOperatorStateAsString := ""
-	if err := wait.Poll(waitPollInterval, waitPollTimeout, func() (bool, error) {
-		clusterOperator, err := operatorClient.Get("cluster", metav1.GetOptions{})
-		if errors.IsNotFound(err) {
-			t.Log("KubeAPIServer/cluster operator does not yet exist.")
-			return false, nil
+	waitNoKeyPollInterval := 15 * time.Second
+	waitNoKeyPollTimeout := 6 * time.Minute
+	waitDuration := 5 * time.Minute
```

Comment:
What are these timeouts? Where do they come from?

Reply:
in some cases, we know that a new key will not be created, for example, running `aescbc` scenario one after another will create only one key. Here the tests wait up to 5 min before checking if resources were migrated.

Reply:
Isn't there a better way to know that? Waiting minutes for nothing to happen is not a good idea IMO.